### ISSUE TITLE
A way to generate a v2 sparse uncompressed fseq for a controller and save it to the local computer

### ISSUE THREAD

**Issue by rledonne** [smeighan/xLights#4025]
Feature request: A way to generate a FSEQ v2 sparse uncompressed for a controller and save it to the local computer so it can be manually transfer to a the controller's SD card.

Background: I have a digquad with the AE+ SD card reader and EsPixelStick firmware. It will often choke on large sequences when uploading from FPP Connect. From my experimenting, it will only play V2 Uncompressed fseq files. xLight's default for FSEQ is V2 zstd, which it will not play, but if you change the system preferences for sequence to V2 Uncompressed, you can save the sequence and manually copy the entire FSEQ from the show folder on to the SD card, and it will be able to play it.

It would be nice, if when trying to use FPP Connect, it could generate the V2 Sparse Uncompressed FSEQ for that controller and save it to the computer, and I can move the file to the SD card manually.

**Comment by cybercop23** [comment]
How many channels are you trying to push? This is most likely due to the esp32 not being able to handle that much traffic coming in. I too use a DQAE+ but only for a prop that's 1k pixels total.
I'd say that you also post this in Espixelstick GH and have Martin look into it.
Frankly this should work and not needing to do it manually. If you need to add have access to an FPP you can manually run the futils and pull out the data for the DQ channels and get yourself a fseq.

**Comment by rledonne** [comment]
It is about 1200 pixels on that controller.
I agree this is not something that needs to be done manually, but since it isn't working, I am trying to find a work around.  
At least one other person on FB has commented that they would appreciate this feature.  Most of the work should be done already since FPP connect can create the controller specific fseq.  I just need a way to trigger that code to happen.
Maybe a lua script or something similar?  I don't know much about what is involved in scripting.

Can you explain more about the futils you mentioned?  I didn't see anything about it in the xLights or FPP manuals.

**Comment by cybercop23** [comment]
It's an FPP utility not xLights, so you'll need access to FPP installed on a PI or similar.
You can just run futils -h and will give you a list of the parameters you can use.
Sorry but by a computer that I can access my fpps

**Comment by dkulp** [comment]
It would be "/opt/fpp/src/fsequtils -h"

**Comment by dartunder** [comment]
Usage: /opt/fpp/src/fsequtils [OPTIONS] FileName.fseq                                                                                                                                           
                                                                                                                                                                                                
  Options:                                                                                                                                                                                      
   -V                - Print version information                                                                                                                                                
   -v                - verbose                                                                                                                                                                  
   -o OUTPUTFILE     - Filename for Output FSEQ                                                                                                                                                 
   -m FSEQFILE       - FSEQ to merge onto the input, ignoring 0                                                                                                                                 
   -M[ FSEQFILE      - FSEQ to merge onto the input, copy 0                                                                                                                                     
   -f #              - FSEQ Version                                                                                                                                                             
   -c (none|zstd|zlib) - Compession type                                                                                                                                                        
   -l #              - Compression level (-99 for default)                                                                                                                                      
   -r (#-# | #+#)    - Channel Range.  Use - to separate start/end channel                                                                                                                      
                            Use + to separate start channel + num channels                                                                                                                      
                       If used before first -m/-M argument, sets a sparse range of output                                                                                                       
                       If used after -m/-M argument, sets a range to read from last merged sequence.                                                                                            
   -n                - No Sparse. -r will only read the range, but the resulting fseq is not sparse.                                                                                            
   -j                - Output the fseq file metadata to json                                                                                                                                    
   -h                - This help output

**Comment by rledonne** [comment]
Can someone give an example of using fsequtils?
When I use **fsequtils -o /home/fpp/media/sequences/sequencenameNEW.fseq -c none -r 5254|8928 /home/fpp/media/sequences/sequencenameORIGNAL.fseq** (sequence names are actual names when used)
I get 
"**-bash: 8928: command not found
terminate called after throwing an instance of 'std::logic_error'
  what():  basic_string::_M_construct null not valid**"

The only time I have got anything is with "**fsequtils -h**"
I get the helptext listed above.

**Comment by dkulp** [comment]
The range would be "-r 5254-8928"

**Comment by rledonne** [comment]
Thank you for correcting the range.  My other issue was I didn't have "\ " for the spaces in the filename.

It would be great if there was a script or something that did this in xLights before uploading them to FPP, then running the utility and copying the new fseqs back down and to the other controller.

**Comment by cbuxton5200** [comment]
Does the "player" FPP instance have the full sequence?  Is that where you would run the fsequtils?

**Comment by cybercop23** [comment]
If you send it with FPPConnect and select UDP All, then it will have all the info.

**Comment by MartinMueller2003** [comment]
> How many channels are you trying to push? This is most likely due to the esp32 not being able to handle that much traffic coming in. I too use a DQAE+ but only for a prop that's 1k pixels total. I'd say that you also post this in Espixelstick GH and have Martin look into it. Frankly this should work and not needing to do it manually. If you need to add have access to an FPP you can manually run the futils and pull out the data for the DQ channels and get yourself a fseq.

We are regularly seeing the FPP Connect sending truncated files via the HTTP upload. I am also seeing that xLights seems to be generating truncated files when the last part of the sequence data has not changes. For example, if the last 20 fames are "All channels Off", the file will have the first all off frame and the rest will be missing. For the moment, the truncated file detector allowed those files to be played if it was only a frame or two worth of data. I will be tightening that up shortly so this needs to be fixed. Either the header reflects the number of frames of data in the file or the file needs to have the additional data in it.

**Comment by DangRock** [comment]
I was going to add a feature request but found this issue... I'm in the same boat... I run 15 ESPixelSticks and 4 DigQuads and I often have transmission issues using FPP Connect. Definitely it happens more on the sticks and DQs that have more pixels, but sometimes, even if a stick is just a few feet from an access point, it'll have a weak signal (figure it's a hardware issue) and the transfers can time out. It takes a LOT of retrying to get them going. It seems the transfer timeout in FPP Connect is fairly low. Many would finish but they timeout right before the transfer is complete. Sometimes, I have to kick off the transfer from FPP Connect, then grab the temp files from the %appdata%\local\temp\ folder and rename them and then send them using the ESPixelStick firmware file transfer interface. It's a very manual process.   

I know it shouldn't be this way, but it is... and one possible solution would be to have an option on FPP connect to output the .fseq files to a selected folder, organized by controller name. Then, we could transfer them either via ESPixelStick interface or just by bringing the flash drives in. That would save a TON of time after mass sequence rebuilds.  

I will look into futils also. I do use FPP as my scheduler. So it sounds like if I do a full fseq transfer to the PI that runs FPP, I can run futils there to split out the sequences by channels?  That could work.

**Comment by MartinMueller2003** [comment]
FP Connect + Latest ESPS V4 allows xLights to use zip files, significantly reducing the time it takes to transfer the updated files. You may not need to go through all of this trouble. Also, it is my understanding that xLights can save the fseq files it is sending to the ESPS devices.

**Comment by MartinMueller2003** [comment]
Correct. After zip files are transfered, you need to do a reboot of the ESPS to get them unzipped.

**Comment by DangRock** [comment]
I'm running a CI of ESPixelStick v4 that's newer than Beta 5, but is not the latest CI available on the web flasher. Would V4 Beta 5 have this capability?  I was running the same version last year (August 24 CI) and don't recall ever seeing zip files. 

I did find the folder where the fseq files are stored when I render. I thought about sending these to the SDs last year but was concerned that the extra channel data (these files would have all channel data, right?) might slow things down on the execution side... is that something I should worry about?  When I had matrixes actually in my sequences, these files were huge, but I've since replaced all matrixes with videos loaded directly on the PIs that run them, so the typical fseq in my render folder is between 10mb and 40mb, most on the smaller side.

**Comment by cybercop23** [comment]
Exactly. That's why you want sparse. It has to be latest xLights and the latest ESPixelStick within the last month or so. use this... https://espixelstickwebflasher.from-ct.com:5000/

**Comment by DangRock** [comment]
Could you elaborate on that a bit?  The FSEQ Version I have set in xLights --> Preferences --> Sequences is "V2 ZSTD (Default)".  Should I change that to "V2 ZSTD/sparse" ? I assume that wouldn't change anything with the FSEQ files generated while rendering IN xLights. FPP Connect definitely uses sparse, and I expect changing it in xLights Preferences won't do anything since I'm not using xSchedule? 

I'll try the updated ESPixelStick firmware again. I'd held off on updating as my experience is the Aug 2024 release has the most stable web interface and the more recent versions have had some issues there, BUT... if the file transfers through FPP Connect are faster, I won't spend nearly as much time in the web interface since I was only in there to transfer the files directly when FPP Connect would time out.

**Comment by cybercop23** [comment]
It needs to be the latest CLI for this ZIP to work. then FPPConnect will discover its ability and change it as such

**Comment by MartinMueller2003** [comment]
You need to use a March 2025 or newer build for zip support to work.

**Comment by DangRock** [comment]
I just tested this out tonight on a DigQuad. Works incredibly well. Initial transfer file (.xlz) was only 134kb and the extracted .fseq (post-reboot) was 14mb. Insane improvement on transfer file size.  Nice work! 

Quick follow-up. I don't have an ESPixelStick handy, to test, but from your comments above, it sounds like the zip transfer works on a DigQuad because it's ESP32-based but not an ESPixelStick, since it's ESP8266?  Not a huge deal since I use the pixel sticks for smaller props, but wanted to ask. 

The only other thing I'll mention as a bug report is that the fseq file date is 12/30/1979 6:00:00 PM. It was correct on the .xlz. Just a note. Tried it twice to see if maybe it was just on the initial file creation (it is persistent). 

Appreciate all the work y'all do.

**Comment by MartinMueller2003** [comment]
ESPixelstick V4 Software executes on over 30 platforms including the dig series devices. It supports both ESP8266 and ESP32. On the ESP32, it supports up to 8 output ports plus I2C pwm outputs.
